src/web/RefreshWorker.ts

```diff
@@ -0,0 +1,129 @@
+import { decodeJwt } from "jose";
+import { renewTokens } from "../core/RefreshTokenGrant";
+import { TokenDetails } from "../core/SessionInformation";
+import { SessionIDB } from "./SessionDatabase";
+
+export enum RefreshMessageTypes {
+    SCHEDULE = 'SCHEDULE',
+    REFRESH = 'REFRESH',
+    STOP = 'STOP',
+    DISCONNECT = 'DISCONNECT',
+    TOKEN_DETAILS = 'TOKEN_DETAILS',
+    ERROR_ON_REFRESH = 'ERROR_ON_REFRESH',
+    EXPIRED = 'EXPIRED'
+}
+
+interface SharedWorker { // to make tsc happy
+    onconnect: (event: MessageEvent) => void;
+}
+declare const self: SharedWorker; // to make tsc happy
+
+// A Set to store all connected ports (tabs)
+const ports: any = new Set();
+let refreshTimeout: any;
+let finalLogoutTimeout: any;
+let timersAreRunning = false;
+let tokenDetails: TokenDetails | undefined = undefined;
+let exp: number | undefined;
+
+self.onconnect = (event: any) => {
```

Comment:
Regarding testability: You could extract the onconnect function to its own file to unit test it. The timeouts can be tested using jest.useFakeTimers()